## chapter 5

### question 2

for k in 1:(n-1)

compute multipliers needs flop: (n-k) 

adjust rows: $2(n-k)^2$

$\sum^{n-1}_1 2(n-k)^2 +  (n-k) $ = $\sum^{n-1}_1 m^2 + m$ = $ {(m-1)(m)(2m-1) \over 6} + {(m)(m-1 \over 2}

$= {2n^3 - 3n^2 +n \over 3} + {n^2 -n \over 2}$

so that the flop is O($2n^3 \over 3$)

### question 3
a)

we have two lower triangular matrix. A and B

let, 0 < i < j < n, $a_{ij}$ = 0, $b_{ij}$ = 0 and we calculate entries in C= A*B, 

$c_{ij} = a_{i1} b_{1j}  + a_{i2}  b_{2j}  + ... + a_{ii}  b_{ij}  ... +a_{ij}  b_{jj}  + ... a_{in}  b_{nj} $

we have know that for $b_{1j},...,b_{j-1,j} = 0$, so $a_{i1} b_{1j}  + a_{i2}  b_{2j}  + ... + a_{ii}  b_{ij}  ... +a_{ij}  b_{jj}$=0

and we know  $a_{ij},...,a_{in} = 0$, so $a_{ij}  b_{jj}  + ... a_{in}  b_{nj}$ = 0

hence they added up, $c_{ij}$ = 0

b)we show this by induction

when n = 2

$$\begin{bmatrix} a & b  \\ 0  &d \end{bmatrix}$$, the inverse is $${1\over ad} \begin{bmatrix} d & -b  \\
0  &a\end{bmatrix}$$, which is also a  upper triangular

$$X_i = \begin{bmatrix} X_{i1} & 0 & .. . & 0\\ 
                        0 & X_{i2} & ... & 0 \\
                        ...  \\ 
                        0 & ... & 0 & X_{ik}\end{bmatrix}$$
                        
$$X_{ik} =\begin{bmatrix} x^t_{i1k}, ... , x^t_{in_{ik}k} \end{bmatrix}  $$                         
                        
$$Z_i = \begin{bmatrix} Z_{i1} & 0 & .. . & 0\\ 
                        0 & Z_{i2} & ... & 0 \\
                        ...  \\ 
                        0 & ... & 0 & Z_{ik}\end{bmatrix}$$
                        
$$Z_{ik} =\begin{bmatrix}z^t_{i1k}, ... , z^t_{in_{ik}k} \end{bmatrix}  $$  

$$\Sigma_i = \begin{bmatrix} \sigma_{i1}I_{n_{ik}} & 0 & .. . & 0\\ 
                        0 & \sigma_{i2}I_{n_{ik}} & ... & 0 \\
                        ...  \\ 
                        0 & ... & 0 & \sigma_{ik}I_{n_{ik}}\end{bmatrix}$$


suppose n = k is true, $A_k$  is an upper trangular , inverse $A_k$ is also anuppter trangular,

let $$A_{k+ 1}  = \begin{bmatrix} A_k & a2  \\ 0  &a3 \end{bmatrix}$$, 

and B is the inverse of $A_{k+ 1}$ and B is 
$$\begin{bmatrix} A^{-1} & b2  \\ b3  & b4 \end{bmatrix}$$

because B is inverse of A, A*B = $I_{n+1}$

so $$ A*B  = \begin{bmatrix} A*A^{-1} & c2  \\ a3b3  &c3 \end{bmatrix} = I$$

hence a3b3 has to be zero for every a3,

so b3 = 0 and  $$ B =\begin{bmatrix} A^{-1} & b2  \\ 0  & b4 \end{bmatrix}$$




C) we show this by induction

when n = 2

$$\begin{bmatrix} a & b  \\ 0  &d \end{bmatrix}$$, the inverse is $${1\over ad} \begin{bmatrix} d & -b  \\
0  &a\end{bmatrix}$$, d and a are positive

suppose n = k is true, $A_k$  is an upper trangular , inverse $A_k$ is also anuppter trangular,

let $$A_{k+ 1}  = \begin{bmatrix} A_k & a2  \\ 0  &a3 \end{bmatrix}$$, 

and B is the inverse of $A_{k+ 1}$ and B is 
$$\begin{bmatrix} A^{-1} & b2  \\ 0  & b4 \end{bmatrix}$$

because B is inverse of A, A*B = $I_{n+1}$

so $$ A*B  = \begin{bmatrix} A*A^{-1} & c2  \\ 0  & a3b4 \end{bmatrix} = I$$

hence a3b4 = 1 and a3 is positive,

so that b4 is positive

C) we show this by induction

when n = 2

$$\begin{bmatrix} 1 & b  \\ 0  &1 \end{bmatrix}$$, the inverse is $$ \begin{bmatrix} 1 & -b  \\
0  &1\end{bmatrix}$$, 

suppose n = k is true, $A_k$  is an upper trangular , inverse $A_k$ is also anuppter trangular,

let $$A_{k+ 1}  = \begin{bmatrix} A_k & a2  \\ 0  &1 \end{bmatrix}$$, 

and B is the inverse of $A_{k+ 1}$ and B is 
$$\begin{bmatrix} A^{-1} & b2  \\ 0  & b4 \end{bmatrix}$$

because B is inverse of A, A*B = $I_{n+1}$

so $$ A*B  = \begin{bmatrix} A*A^{-1} & c2  \\ 0  & b4 \end{bmatrix} = I$$

hence b4 = 1 and a3 is positive,

$$\begin{bmatrix} A^{-1} & b2  \\ 0  & 1 \end{bmatrix}$$ with all diagnoal entris are 1

### question 4

we want to show that an orthogonal upper-triangular matrix is diagonal matrix

we can prove this by incudction

when n = 2

$$\begin{bmatrix} a & b  \\ 0  &d \end{bmatrix}$$

b has to be zero, otherwise $a_1^t a_2 = ab \neq 0$  and contradict with matrix is an orthogonal matrix

suppose for n, $A_n$ is  diagonal

for  n + 1, $$B_{n+1} =\begin{bmatrix} A & b_n  \\ 0  &d \end{bmatrix}$$

soppose there is a non-zero emelemt in $b_n$ is $b_{in}$

$a_i * b_n$  = $a_{1i}a_{1n} + ...+ a_{ii}a_{in}.. + a_{ni}a_{nn}$

becasue it is upper trangular matrix, so $a_{ii}$ is not equal to zero, $a_{ii}a_{in} \neq 0$,

because A is diagonial, so that $a_{ij} = 0$ for $ i \neq j$

hence $a_i * b_n = 0 + ... + a_{ii}a_{in}  \neq 0 $, which contradicts with B is an orthogonal matrix.

hence we showed that an orthogonal upper-triangular matrix is diagonal is also true for n+1.

the statement has be proved


### question 7 Find by hand the Cholesky decomposition of the matrix

$$A =\begin{bmatrix} 2 & -2  \\ -2  &5 \end{bmatrix}
= \begin{bmatrix} L11 & 0  \\ L  &L22 \end{bmatrix}
\begin{bmatrix} L11 & L  \\ 0  &L22 \end{bmatrix}$$

2 = $L11^2$, L11= $\sqrt(2)$

-2 = L11L, L= $-\sqrt(2)$ 

5 = $L^2 + L22^2$  L22= $\sqrt(3)$ 

therefore 

$$A
= \begin{bmatrix} \sqrt(2) & 0  \\ -\sqrt(2)  &\sqrt(3) \end{bmatrix}
\begin{bmatrix} \sqrt(2) & -\sqrt(2)  \\ 0  &\sqrt(3) \end{bmatrix}$$

### question 8  Show that the matrices are both valid Cholesky-like decompositions of the positive semidefinite matrix

Cholesky Decomposition $ LL^t = A$


In [20]:
B = [1 0 0
2 0 0
2 3 2]
B * transpose(B)

3×3 Matrix{Int64}:
 1  2   2
 2  4   4
 2  4  17

we shown that B is lower-triangular matrix such that $BB^t=A$, so B is Cholesky-like decompositions of A

In [21]:
C = [1 0 0
2 0 0 
2 0 13^0.5]
C * transpose(C)


3×3 Matrix{Float64}:
 1.0  2.0   2.0
 2.0  4.0   4.0
 2.0  4.0  17.0

so we shown that B is lower-triangular matrix such that $CC^t=A$, so C is Cholesky-like decompositions of A

### question 9 we want to show that the Cholesky decomposition of band matrix is also a mabd matrix

we want to show this by induction

for n = 2, 

$$A  =\begin{bmatrix} a1 & a2  \\ a3 & a4 \end{bmatrix}
= \begin{bmatrix} L11 & 0  \\ L  &L22 \end{bmatrix}
\begin{bmatrix} L11 & L  \\ 0  &L22 \end{bmatrix}$$

if d = 0, then $$A  =\begin{bmatrix} a1 & 0  \\ 0 & a4 \end{bmatrix} $$

L*L11 = 0, L = 0,

$$L=  \begin{bmatrix} L11 & 0  \\ 0  &L22 \end{bmatrix}$$

so it is true for n = 2


then we suppose that if $A_n$ is band matrix, then its decomposition is also a band matrix, and we show that it is true for $A_{n+1}$

A is banded by d, so that  $$A_{n+1}  =\begin{bmatrix} a11 & a12  & ...& a1(d+1) & 0 &... & 0\\ 
a21 \\
...\\
a(d+1)1&&A_{n}\\
0\\
...\\
0 \end{bmatrix}$$

$$= \begin{bmatrix} l11 & 0  & ...&0 & 0 &... & 0\\ 
l21\\
...\\l(d+1)1&&L_{n}\\
l(d+2)1\\
...\\
ln1 \end{bmatrix}
 \begin{bmatrix} l11 & l21  & ...&l(d+1)1&l(d+2)1  &... & ln1\\ 
0\\
...\\0   &&& &L_{n}\\
0\\
...\\
0 \end{bmatrix}$$

we want to solve equation $a = l11l$, which is equavalent to
$$\begin{bmatrix} a21 \\
...\\
a(d+1)1\\
0\\
...\\
0 \end{bmatrix} = l11
\begin{bmatrix} l21 \\
...\\
l(d+1)1\\
l(d+2)1\\
...\\
ln1 \end{bmatrix} $$

therefore we can we that li1= 0 fot i-1>d.

so that the Cholesky decomposition  is 

$$= \begin{bmatrix} l11 & 0  & ...&0 & 0 &... & 0\\ 
l21\\
...\\l(d+1)1&&L_{n}\\
0\\
...\\
0 \end{bmatrix}$$

we know that for $L_n$, lij= 0 if |j-i|>d and we also shown that fot the first column, lij= 0 if |j-i|>d, hence we proved that $L_{n+1}$ is also a band matrix.

then the statement has been proved

### question 11  Show that inversion of an arbitrary square matrix B can be reduced to inversion of a positive definite matrix

we know$B^{-1}B= I$

multiply $B^T$ each side, then we get $B^{-1}BB^T= B^T$

then multiply $(BB^T)^{-1}$ each side,  then we get $B^{-1}= B^T(BB^T)^{-1}$

since $(BB^T)$ is positive definite matrix, hence we shown that  inversion of an arbitrary square matrix B can be reduced to inversion of a positive definite matrix



## chapter 6

### question 1

$ f(x) = e^x - x - 2$

$ f'(x) = e^x - 1$

$x_{n+1} = x_n - {e^x - x - 2 \over e^x - 1}$

In [2]:
function newton(f::Function, fp::Function, x::Real, tol::Real)
(xold, xnew) = (x, zero(x))
for iteration = 1:100
xnew = xold - f(xold) / fp(xold)
if abs(f(xnew)) < tol
return (xnew, iteration)
end
xold = xnew
end
return (xnew, 100)
end

f(x) = exp(x) - x - 2
fp(x) = exp(x) - 1
(x, iterations) = newton(f, fp, 2.5, 1e-14)

(1.1461932206205825, 7)

so that the answer of the equation is 1.1461932206205825, then we check with `fzero` function

In [3]:
import Pkg; Pkg.add("Roots")
using Roots
f(x) = exp(x) - x - 2
fzero(f,1,2)


   Resolving package versions...
  No Changes to `~/.julia/environments/v1.8/Project.toml`
  No Changes to `~/.julia/environments/v1.8/Manifest.toml`


1.1461932206205825

the results shows thats same

### question 2

first we use newton method

$f(x) = we^w - x$

$f'(x) = we^w + e^w$

$w_{n + 1} = w - {(we^w -x) \over (e^w + we^w)}
= {(we^w + w^2 e^w - we^w +x)\over (e^w + we^w)}
= {( w^2 e^w  +x)\over (e^w + we^w)}
= {w( w e^w  +{x\over w})\over (e^w + we^w)}
= {w( w   +{x\over we^w})\over (1 + w)}$

therefore we shows the iterates

we know $w_{n + 1}= {w_n( w_n   +{x\over w_ne^w_n})\over (1 + w_n)}$

which is equavalent with ${w_{n + 1} \over w_n} = {( w_n   +{x\over w_ne^{w_n}})\over (1 + w_n)}$

for $x <  w_ne^{w_n}$,$ {x\over w_ne^w_n} < 1$,$ {( w_n   +{x\over w_ne^{w_n}}) < (1 + w_n)}$, which shows that 
${( w_n   +{x\over w_ne^{w_n}})\over (1 + w_n)} < 1$ and then ${w_{n + 1} \over w_n} < 1$, ${w_{n + 1} < w_n}$

for $x >  w_ne^{w_n}$,$ {x\over w_ne^w_n} > 1$,$ {( w_n   +{x\over w_ne^{w_n}}) > (1 + w_n)}$, which shows that 
${( w_n   +{x\over w_ne^{w_n}})\over (1 + w_n)} > 1$ and then ${w_{n + 1} \over w_n} > 1$, ${w_{n + 1} >w_n}$

In [47]:
function lambert(x::Real)
y = exp(1) # initional w
for n in 1:50
z = exp(y)
y = y*(y + x/(y*z))/(y+1) #calculate iterates
end
return y
end

lambert (generic function with 1 method)

In [50]:
lambert(5)


1.3267246652422002

In [49]:
lambert(exp(1))

1.0

### question 4

we know from the question that 
$x_{n + 1} -x_n = -{f(x_n) \over f'(x_n) + 0.5 f''(x_n)(x - x_n) }$

and $x_{n + 1} -x_n = -{ f(x_n) \over f'(x_n)}$

so $x_{n + 1} -x_n = -{f(x_n) \over f'(x_n) - 0.5 f''(x_n){ f(x_n) \over f'(x_n)}}$

 $x_{n + 1} -x_n = -{2f(x_n)f'(x_n) \over 2f'(x_n)f'(x_n) -  f''(x_n){ f(x_n) }}$

the iterates of Lambert’s equation is 

$w_{j + 1} = w - {we^w -x \over {e^ww + e^w - {(w+2)(we^w-x) \over 2w+2} }   }$

Halley convergence faster so it need less iterates, but Halley is more complex, need more calculation.

### question 5

$f'(x) = 2 (sin({2 \over x})x -cos({2 \over x}))$

$x_{n+1} = x_n -{x + x^2sin({2 \over x}) \over 2 (sin({2 \over x})x -cos({2 \over x}) ) + 1}$

$x_{n+1} - x_n = {{x } \over 2 (sin({2 \over x})x -cos({2 \over x}) ) + 1} ={1 \over 2 (sin({2 \over x}) -{1\over x}cos({2 \over x}) ) + {1\over x}} $

for x is close to 0, $1 \over x$ is close to positive infinity, and $2(sint -t cost) + t$ is from negative infinity and positive infinity, therefore $x_{n+1} - x_n$ is from negative infinity and positive infinity, which is not convergence, hence it is failed when x = 0

### question 6

$f(x) = x + x^{4\over3}$

$f'(x) = 1 + {4 \over 3}x^{1\over3}$

$x_{n+1}=x_{n }- {x + x^{4\over3} \over 1 + {4 \over 3}x^{1\over3}}
= {{ x^{4\over3} \over 3} \over 1 + {4 \over 3}x^{1\over3}}$


${x_{n+1} \over x } =  {x^{1\over3} \over {4 \over 3}x^{1\over3} + 1}$, we can see from this equation that $x_{1}$ is decreasing

${x_{n+1} \over x^{4\over3} } =  {{1\over 3} \over {4 \over 3}x^{1\over3} + 1}$, $x_0$ is close to 0, and ${x_n}$ is decreasing,  so that ${4 \over 3}x^{1\over3} $ is close to 0,  ${{4 \over 3}x^{1\over3} + 1}$ is close to 1, hence we shown that $lim{x_{n+1} \over x^{4\over3} } = {1 \over 3}$
